pkg/networkserver/downlink.go

```diff
@@ -967,18 +987,31 @@ func (ns *NetworkServer) scheduleDownlinkByPaths(ctx context.Context, req *sched
 			"transmit_at", transmitAt,
 			"absolute_time", ttnpb.StdTime(req.TxRequest.AbsoluteTime),
 		)).Debug("Scheduled downlink")
-		queuedEvents = append(queuedEvents, events.Builders(append([]events.Builder{
-			successEvent.With(
-				events.WithData(&ttnpb.ScheduleDownlinkResponse{
-					Delay: ttnpb.ProtoDurationPtr(delay),
-				}),
-			),
-		}, []events.Builder(req.DownlinkEvents)...)).New(ctx, eventIDOpt)...)
+		queuedEvents = append(queuedEvents, successEvent.With(
+			events.WithData(&ttnpb.ScheduleDownlinkResponse{
+				Delay: ttnpb.ProtoDurationPtr(delay),
+			}),
+		).New(ctx, eventIDOpt))
 		registerSuccess(ctx)
-		return &scheduledDownlink{
+		// Mark the group as having a successful transmission, in order to avoid
+		// subsequent re-attempts.
+		groupDone[a.groupIndex] = struct{}{}
+		// Report to the upper layer only the latest (chronological) transmission
+		// for book keeping purposes (such as transmission times).
+		if latestScheduledDownlink != nil && latestScheduledDownlink.TransmitAt.Sub(transmitAt) > 0 {
+			continue
+		}
+		latestScheduledDownlink = &scheduledDownlink{
 			Message:    down,
 			TransmitAt: transmitAt,
-		}, queuedEvents, nil
+		}
+		downlinkEvents = req.DownlinkEvents.New(ctx, eventIDOpt)
+	}
+	// NOTE: A singular successful scheduling attempt will render the downlink as a whole as
+	// being successfully transmitted, even if certain groups did not have a successful scheduling
+	// attempt.
```

Comment:
I don't understand this comment. What is a singular scheduling attempt?

Reply:
This is related to the follow section of the 1.0.4 spec:
<img width="931" alt="image" src="https://user-images.githubusercontent.com/36161392/172684707-6dd3d7b5-981c-42bb-900a-27b1ec9dcb45.png">

I've updated the wording in 7cf7e5b6b5e59ea041106448dcf13987997e873f - let me know if this is more clear.

Reply:
Right, clear now.